Snapshot to chainspec command

ControlCplusControlV opened the pull request.

ControlCplusControlV commented:
Moved to PR so I can see everything to clean up easier, mostly working but still sorting out a cleaner test and potential state_config issue

rakita commented:
I did a little bit of refactoring of the main config: https://github.com/FuelLabs/fuel-core/pull/440 just a heads up as there will be conflicts.

ControlCplusControlV commented:
Ok, going to hold off requesting review until those are merged then

ControlCplusControlV commented:
Converting to draft while I removed feature flags as requested

ControlCplusControlV commented:
Closes #75

leviathanbeak commented on fuel-core/src/config/chain_config.rs:
```diff
@@ -121,6 +120,16 @@ pub struct StateConfig {
+impl StateConfig {
+    pub fn generate_state_config(db: Database) -> Self {
+        StateConfig {
+            coins: db.get_coin_config().unwrap(),
```
are these things always assumed to work or should we be more careful with `unwraps` here ?

ControlCplusControlV replied:
Moving this to `unwrap_or_default`, these should always work and properly return a `None` value if nothing exists, but moving to `unwrap_or_default` so an error in one of them doesn't ruin the generation of the entire config

leviathanbeak commented on fuel-core/src/database/coin.rs:
```diff
@@ -120,4 +123,36 @@ impl Database {
+        let configs = self
+            .iter_all::<Vec<u8>, Word>(COIN, None, None, None)
+            .map(|raw_coin| -> CoinConfig {
+                let coin = raw_coin.unwrap();
```
few `unwraps` here as well - would it be better that we rather return an `error` instead of panicking here?

ControlCplusControlV replied:
I initialing had trouble unwrapping inside of that closure because I wasn't returning an error, but I agree that it is dangerous and will adjust this to properly propagate errors to the top

leviathanbeak commented on fuel-core/src/database/contracts.rs:
```diff
@@ -67,6 +69,46 @@ impl Database {
+                let contract_id =
+                    ContractId::new(raw_contract_id.unwrap().0[..32].try_into().unwrap());
+
+                // Working?
```
`Working?` comments here - what are they for :D

leviathanbeak reviewed (dismissed):
I think there's a lot of `unwraps` everywhere, in methods that actually return `Result<_, anyhow::Error>` so it would be better to cast those errors into `anyhow::Error` and return them rather than potentially `unwrap()` panicking in the middle of the execution.

ControlCplusControlV commented:
> I think there's a lot of `unwraps` everywhere, in methods that actually return `Result<_, anyhow::Error>` so it would be better to cast those errors into `anyhow::Error` and return them rather than potentially `unwrap()` panicking in the middle of the execution.

I agree, I moved both coin and contract config generation to now `unwrap` much smarter and propagate the error upward if one does occur. I also dropped in better error handling in `state_config` generation by falling back to a default value in an error occurs so the rest of the config can still be exported in the event there is in error in one part of it.

Voxelot commented on fuel-core/src/args.rs:
```diff
@@ -12,14 +22,17 @@ lazy_static::lazy_static! {
 )]
 pub struct Opt {
+    #[clap(subcommand)]
+    pub _snapshot: Option<Snapshot>,
```
[nit] any reason for the `_` prefix here?

Voxelot commented on fuel-core/src/args.rs:
```diff
@@ -137,3 +178,34 @@ impl Opt {
+
+    #[cfg(feature = "rocksdb")]
+    {
+        let db = Database::open(&path)?;
```
```suggestion
        let db = Database::open(&path).context(|| format!("failed to open database at path {}", path))?;
```

Voxelot commented on fuel-core/src/args.rs:
```diff
@@ -137,3 +178,34 @@ impl Opt {
+            transaction_parameters: _config.transaction_parameters,
+        };
+
+        let serialized = serde_json::to_string(&chain_conf).unwrap();
+        let mut stdout = io::stdout().lock();
+
+        stdout.write_all(serialized.as_bytes())?;
```
1. No need to unwrap here when we have the convenience of anyhow available
2. Consider using streaming serialization instead of copying the whole string into memory

```suggestion
        let mut stdout = io::stdout().lock();
        let serialized = serde_json::to_writer(stdout, &chain_conf).context("failed to dump snapshot to JSON")?;
```

ControlCplusControlV replied:
I added this in but not commit suggestion because it actually further can have some changes such to make this look like
```rust
        let stdout = io::stdout().lock();
        serde_json::to_writer(stdout, &chain_conf).context("failed to dump snapshot to JSON")?;
        ```

Voxelot commented on fuel-core/src/database/coin.rs:
```diff
@@ -120,4 +123,36 @@ impl Database {
+        let configs = self
+            .iter_all::<Vec<u8>, Word>(COIN, None, None, None)
+            .map(|raw_coin| -> Result<CoinConfig, anyhow::Error> {
+                let coin = raw_coin.unwrap();
```
is unwrap needed here?

Voxelot commented on fuel-core/src/database/coin.rs:
```diff
@@ -120,4 +123,36 @@ impl Database {
+
+                let ref_coin = Storage::<UtxoId, CoinModel>::get(self, &tx_id)?;
+
+                let ref_coin = ref_coin.unwrap();
```
more unwrap

ControlCplusControlV replied:
This is an option Unwrap not a result unwrap, so should be safe right?

Voxelot replied:
Why do we need to fetch the coin by ID if we've already unwrapped it from `raw_coin` above?

Voxelot commented on fuel-core/src/database/contracts.rs:
```diff
@@ -67,6 +69,40 @@ impl Database {
+                    .0;
+
+                let state: Option<Vec<(Bytes32, Bytes32)>> =
+                    self.get(contract_id.as_ref(), CONTRACTS_STATE)?;
```
`get` only fetches a single key, we need to use `iter_all` to collect _all_ the state of the contract. Also these columns use a `MultiKey`, where contract id is just the prefix to the actual state key.

Voxelot commented on fuel-tests/tests/snapshot.rs:
```diff
@@ -0,0 +1,74 @@
+        contracts: Some(vec![ContractConfig {
+            code: vec![8; 32],
+            salt: Salt::new([9; 32]),
+            state: None,
+            balances: None,
```
test should cover state and balances as well, as it would reveal current problems in the implementation

ControlCplusControlV commented on fuel-core/src/args.rs:
```diff
@@ -137,3 +178,34 @@ impl Opt {
     }
 }
+
+pub fn dump_snapshot(path: PathBuf, _config: ChainConfig) -> anyhow::Result<()> {
```
config is unused if rocksdb isn't enabled because a snapshot couldn't be taken, and so it's never used. I got around this with path by printing it in the error, would you like me to do the same with config?

Voxelot replied:
Ah makes sense, I think it's fine as is then.

rakita replied:
Usually, you specify two functions one with an error and one without it and use cfg to enable/disable them.

Voxelot commented on fuel-tests/Cargo.toml:
```diff
@@ -32,10 +35,12 @@ fuel-txpool = { path = "../fuel-txpool" }
 futures = "0.3"
 insta = "1.8"
 itertools = "0.10"
+portpicker = "0.1.1"
```
This doesn't seem necessary now that we removed full integ testing of the CLI

ControlCplusControlV replied:
will remove

Voxelot commented on fuel-tests/Cargo.toml:
```diff
@@ -16,14 +16,17 @@ name = "integration_tests"
 [dependencies]
+async-std = "1.12"
 chrono = { version = "0.4", features = ["serde"] }
+escargot = "0.5.7"
```
Also not needed anymore

ControlCplusControlV replied:
will remove

rakita commented on fuel-core/src/args.rs:
```diff
@@ -12,14 +19,17 @@ lazy_static::lazy_static! {
     version,
     rename_all = "kebab-case"
 )]
 pub struct Opt {
```
https://github.com/FuelLabs/fuel-core/issues/457

ControlCplusControlV replied:
I was wondering if it was worth re-vamping the cli structure with this PR but wanted to keep changes minimal. I definitely agree though because this implementation is a bit clunky

rakita commented on fuel-core/src/args.rs:
```diff
@@ -64,8 +74,35 @@ pub struct Opt {
+#[derive(clap::Subcommand, Clone, Debug)]
+pub enum Snapshot {
+    Snapshot(SnapshotCommand),
+}
```
[nit] This probably deserved it's own mod, as in `args/snapshot.rs`, can be put under effort of refactoring cli subcommand: https://github.com/FuelLabs/fuel-core/issues/457

Voxelot commented on fuel-core/src/args/run.rs:
```diff
@@ -0,0 +1,149 @@
+
+impl RunCommand {
+    pub fn get_config(self) -> io::Result<Config> {
+        let filter = match env::var_os(LOG_FILTER) {
+            Some(_) => EnvFilter::try_from_default_env().expect("Invalid `RUST_LOG` provided"),
+            None => EnvFilter::new("info"),
+        };
+
+        let human_logging = env::var_os(HUMAN_LOGGING)
+            .map(|s| {
+                bool::from_str(s.to_str().unwrap())
+                    .expect("Expected `true` or `false` to be provided for `HUMAN_LOGGING`")
+            })
+            .unwrap_or(true);
+
+        let sub = tracing_subscriber::fmt::Subscriber::builder()
+            .with_writer(std::io::stderr)
+            .with_env_filter(filter);
+
+        if human_logging {
+            // use pretty logs
+            sub.with_ansi(true)
+                .with_level(true)
+                .with_line_number(true)
+                .init();
+        } else {
+            // use machine parseable structured logs
+            sub
+                // disable terminal colors
+                .with_ansi(false)
+                .with_level(true)
+                .with_line_number(true)
+                // use json
+                .json()
+                .init();
+        }
```
As this logging setup could be used for both snapshot and running, it should be moved to the central `run_cli` rather than only happening when a node is run.

ControlCplusControlV replied:
Moved it into the outer `args.rs` so it should be accessible for subcommands, and starts regardless of the subcommand being run

Voxelot commented on fuel-tests/tests/snapshot.rs:
```diff
@@ -0,0 +1,95 @@
+
+    let contract_config = &state_conf.contracts.unwrap()[1];
+
+    assert!(contract_config.balances.is_some());
+    assert!(contract_config.state.is_some());
```
We should have assertions that the state key, state values, and balances are correct, and not just set to any random value. Perhaps we should split this test up across the various areas of concern, such as a test for each of these:
- contract deployments (eg. bytecode, salt)
- contract balances
- contract state
- coins
- block height

These tests should verify the exact expected values.

ControlCplusControlV replied:
Ok, will expand this test. Sorry didn't see this until I had just re-requested review in the 2nd commit

ControlCplusControlV commented:
I removed the expensive and ugly double collection, and then remove the extra get within the balances state collection (fixing what @Voxelot had asked in using all of the values provided in `iter_all`)

Voxelot commented on fuel-core/src/config/chain_config.rs:
```diff
@@ -121,6 +120,16 @@ pub struct StateConfig {
 }
 
+impl StateConfig {
+    pub fn generate_state_config(db: Database) -> Self {
+        StateConfig {
+            coins: db.get_coin_config().unwrap_or_default(),
+            contracts: db.get_contract_config().unwrap_or_default(),
+            height: db.get_block_height().unwrap_or_default(),
```
This method should propagate the errors from the database methods without unwrap. Otherwise, the snapshot could silently fail while still providing partially valid results without the user knowing.

```suggestion
    pub fn generate_state_config(db: Database) -> anyhow::Result<Self> {
        StateConfig {
            coins: db.get_coin_config()?,
            contracts: db.get_contract_config()?,
            height: db.get_block_height()?,
```

ControlCplusControlV replied:
I initially did this so corruption in one column would allow the rest to be saved, but I see how this silent failure could be more harmful

Voxelot replied:
This could be enabled in the future using a `continue-on-error` flag, however, any errors should be reported via `error!` or `warn!` before `unwrap_or_default` so the user is aware that their snapshot may come with some issues.
